### ISSUE TITLE
The position of the extracted image is incorrect

### ISSUE THREAD

**Issue by edisonwd** [pymupdf/RAG#189]
The reproduced code is as follows:

```python
import hashlib
import os
from pathlib import Path
from tempfile import TemporaryDirectory
from typing import Optional, Dict
from urllib.parse import urlparse

import pymupdf4llm
import requests


def is_valid_url(url: str) -> bool:
    """Check if the url is valid."""
    parsed = urlparse(url)
    return bool(parsed.netloc) and bool(parsed.scheme)


def handle_remote_file(file_path: str, temp_dir: TemporaryDirectory, headers: Optional[Dict] = None):
    # If the file is a web path, download it to a temporary file, and use that
    if not os.path.isfile(file_path) and is_valid_url(file_path):
        web_path = file_path
        r = requests.get(web_path, headers=headers)
        if r.status_code != 200:
            raise ValueError(
                "Check the url of your file; returned status code %s"
                % r.status_code
            )
        _, suffix = os.path.splitext(file_path)
        suffix = suffix.split("?")[0]

        md5_hash = hashlib.md5(file_path.encode('utf-8')).hexdigest()
        temp_pdf = os.path.join(temp_dir.name, f"tmp_{md5_hash}{suffix}")
        with open(temp_pdf, mode="wb") as f:
            f.write(r.content)
        file_path = str(temp_pdf)
    elif not os.path.isfile(file_path):
        raise ValueError("File path %s is not a valid file or url" % file_path)

    return file_path


def pdf_to_markdown(file_path: str,
                    temp_dir: TemporaryDirectory,
                    md_file_path: str = None,
                    image_path: str = None,
                    write_images: bool = True,
                    image_format: str = "png",
                    dpi: int = 200):
    file_path = handle_remote_file(file_path, temp_dir=temp_dir)

    if md_file_path is None:
        filename, ext = os.path.splitext(os.path.basename(file_path))
        md_file_path = os.path.join(temp_dir.name, f"{filename}.md")

    if image_path is None:
        image_path = os.path.join(os.path.dirname(md_file_path), f"images")

    md_text = pymupdf4llm.to_markdown(doc=file_path,
                                      write_images=write_images,
                                      image_path=image_path,
                                      image_format=image_format,
                                      dpi=dpi,
                                      )

    # 写入Markdown文件
    Path(md_file_path).write_bytes(md_text.encode())
    print(f"Markdown file path: {md_file_path}")

    return md_file_path


if __name__ == '__main__':
    file_path = "https://arxiv.org/pdf/2411.07264"
    md_file_path = pdf_to_markdown(file_path=file_path,
                                   temp_dir=TemporaryDirectory(),
                                   md_file_path="test2.md",
                                   )
    print(md_file_path)

```
When I used the above code to convert the PDF to markdown format, I found that the position of the extracted image was incorrect.

The image position in the converted markdown file is:
![image](https://github.com/user-attachments/assets/c163b964-2e71-431b-a6a3-732d9e958b77)

The expected image position is:
![image](https://github.com/user-attachments/assets/6fd660c0-386c-46dc-a8fd-4335ed8fe6d4)

**Comment by adamcvolz** [comment]
Yeah, experiencing this as well, it's because they are processing images separately from text and just appending to the end of the markdown text, unless the image comes before any text block, in which it will maintain the top of the page. Looking for a solution myself, hoping metadata of my fitz document or something can help me insert them back, but a fix would be nice

**Comment by philipplelidis** [comment]
Same issue.

**Comment by neverlatetolearn0** [comment]
Add the following code to the get_page_output function
Place the following code after parms.vg_clusters = dict((i, r) for i, r in enumerate(parms.vg_clusters0)) and comment the original text_rects.

        def split_text_rects(text_rects, avoid_rects):
            final_rects = []
            for rect in text_rects:
                overlapping = [avoid for avoid in avoid_rects if rect.intersects(avoid)]

                if not overlapping:
                    final_rects.append(rect)  
                    continue
                overlapping.sort(key=lambda r: r.y0)
）
                split_points = [rect.y0]
                for avoid in overlapping:
                    split_points.extend([avoid.y0, avoid.y1])  
                split_points.append(rect.y1)  

 
                for i in range(0, len(split_points) - 1, 2):
                    y0, y1 = split_points[i], split_points[i + 1]
                    if y1 - y0 > 1:  
                        sub_rect = pymupdf.Rect(rect.x0, y0, rect.x1, y1)
                        final_rects.append(sub_rect)

            return final_rects


        text_rects = column_boxes(
            parms.page,
            paths=parms.actual_paths,
            no_image_text=True,
            textpage=parms.textpage,
            avoid=parms.tab_rects0 + parms.vg_clusters0,
            footer_margin=footer_margin,
            header_margin=header_margin,
        )

        avoid_rects = parms.tab_rects0 + parms.vg_clusters0

        text_rects = split_text_rects(text_rects, avoid_rects)
        text_rects.sort(key=lambda r: (r.y0, r.x0))

**Comment by JorjMcKie** [comment]
Please provide the PDF for reproducing the problem.
Also, the new version 0.0.19 fixes many issues. While this specific problem might still persist, all future discussion must be based on that version.

**Comment by philipplelidis** [comment]
> Please provide the PDF for reproducing the problem. Also, the new version 0.0.19 fixes many issues. While this specific problem might still persist, all future discussion must be based on that version.

The problem persists with 0.0.19 with the following document.

[Operating_Manual_PlanTeam-SPACE_first20.pdf](https://github.com/user-attachments/files/19565470/Operating_Manual_PlanTeam-SPACE_first20.pdf)

**Comment by JorjMcKie** [comment]
> > Please provide the PDF for reproducing the problem. Also, the new version 0.0.19 fixes many issues. While this specific problem might still persist, all future discussion must be based on that version.
> 
> The problem persists with 0.0.19 with the following document.
> 
> [Operating_Manual_PlanTeam-SPACE_first20.pdf](https://github.com/user-attachments/files/19565470/Operating_Manual_PlanTeam-SPACE_first20.pdf)

Thanks for the file. Very helpful for making more changes.
The file is now correctly processed.

**Comment by philipplelidis** [comment]
> > > Please provide the PDF for reproducing the problem. Also, the new version 0.0.19 fixes many issues. While this specific problem might still persist, all future discussion must be based on that version.
> > 
> > 
> > The problem persists with 0.0.19 with the following document.
> > [Operating_Manual_PlanTeam-SPACE_first20.pdf](https://github.com/user-attachments/files/19565470/Operating_Manual_PlanTeam-SPACE_first20.pdf)
> 
> Thanks for the file. Very helpful for making more changes. The file is now correctly processed. The fix will be published with the next version.

Nice! Good to here :) The Issue with images in tables can also be reproduced with this file.

[https://github.com/pymupdf/RAG/issues/21](url)

**Comment by JorjMcKie** [comment]
> The Issue with images in tables can also be reproduced with this file.

Yes, I have seen it. But we won't do anything in that direction in the foreseeable time. There's lots of stuff with a higher priority.
